### PR TITLE
Upgrade tgi to 2.3.1

### PR DESCRIPTION
**Pull request by yuanwu2017**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by mandy-li** [comment]
@yuanwu2017 , for the testing, pls add the torch.compile model to your test cases.

File: server/text_generation_server/models/causal_lm.py [hunk 1]
```diff
@@ -1,499 +1,628 @@
...
 tracer = trace.get_tracer(__name__)
 
+# MAX_TOTAL_TOKENS = int(os.environ.get('MAX_TOTAL_TOKENS', 2048))
+# BATCH_BUCKET_SIZE = int(os.environ.get('BATCH_BUCKET_SIZE', 8))
+# PAD_SEQUENCE_TO_MULTIPLE_OF = int(os.environ.get('PAD_SEQUENCE_TO_MULTIPLE_OF', 128))
+# PREFILL_BATCH_BUCKET_SIZE = int(os.environ.get('PREFILL_BATCH_BUCKET_SIZE', 4))
+# CHUNK_SIZES = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048]
+# LAZY_MODE = int(os.environ.get('PT_HPU_LAZY_MODE', 1))
+MAX_TOTAL_TOKENS = int(os.environ.get('MAX_TOTAL_TOKENS', 8192))
+MAX_BATCH_TOTAL_TOKENS = int(os.environ.get('MAX_BATCH_TOTAL_TOKENS', 65536))
```
Comment (tthakkal) [inline server/text_generation_server/models/causal_lm.py]: @yuanwu2017 looks like you are not reading param `--max-batch-total-tokens` coming from command line, instead always setting it to `MAX_BATCH_TOTAL_TOKENS=65536` , its causing out of memory at "Decode warmup with bigger batch_size"

File: server/text_generation_server/models/causal_lm.py [hunk 2]
```diff
@@ -1160,35 +1203,127 @@ def generate_token(
...
+        batch_size = max_prefill_batch_size * 2
+        # Decode warmup with bigger batch_size
+        try:
+            if DECODE_WARMUP_BATCH_SIZE_LIST[-1] < max_decode_batch_size and batch_size <= max_decode_batch_size:
+                batches = []
+                for i in range(int(batch_size/max_prefill_batch_size)) :
+                    batch = self.generate_warmup_batch(request, PREFILL_WARMUP_SEQLEN_LIST[0], DECODE_WARMUP_BATCH_SIZE_LIST[-1], is_warmup)
+                    _, prefill_batch, _ = self.generate_token([batch], is_warmup)
+                    batches.append(prefill_batch)
+                while batch_size <= max_decode_batch_size:
```
Comment (tthakkal) [inline server/text_generation_server/models/causal_lm.py]: Warming up multiple decode batches causing out of memory. for example below command crashes in this loop with out of memory which used to work in last release.

```
docker run -it --rm -p 8056:80 \
    --runtime=habana \
        -v /data2/models:/data \
    -v /data2/models:/data/hub \
    -v /data2/models:/root/.cache/huggingface/hub \
    -e PT_HPU_ENABLE_LAZY_COLLECTIVES=true \
   -e HABANA_VISIBLE_DEVICES=1 \
   -e MAX_TOTAL_TOKENS=2048 \
   -e PREFILL_BATCH_BUCKET_SIZE=2 \
   -e BATCH_BUCKET_SIZE=32 \
   -e PAD_SEQUENCE_TO_MULTIPLE_OF=256 \
    -e OMPI_MCA_btl_vader_single_copy_mechanism=none \
    -e ENABLE_HPU_GRAPH=true \
    -e LIMIT_HPU_GRAPH=true \
    -e USE_FLASH_ATTENTION=true \
    -e FLASH_ATTENTION_RECOMPUTE=true \
    --entrypoint bash \
    --cap-add=sys_nice --ipc=host <PR_225_image>
   --model-id meta-llama/Llama-2-7b-chat-hf \
   --max-input-length 1024 \
   --max-total-tokens 2048 \
   --max-batch-prefill-tokens 2048 \
   --max-batch-total-tokens 65536 \
   --max-waiting-tokens 7 \
   --waiting-served-ratio 1.2 \
   --max-concurrent-requests 64
```

File: server/text_generation_server/models/causal_lm.py (hunk 1 above)
Comment (yuanwu2017) [reply server/text_generation_server/models/causal_lm.py]: It is only a draft. I am running the performance benchmark and fix some issues I found.

File: server/text_generation_server/models/causal_lm.py (hunk 2 above)
Comment (yuanwu2017) [reply server/text_generation_server/models/causal_lm.py]: Ok. I will fix it.

**Comment by yuanwu2017** [comment]
> @yuanwu2017 , for the testing, pls add the torch.compile model to your test cases.
Ok.

File: server/text_generation_server/models/causal_lm.py (hunk 2 above)
Comment (yuanwu2017) [reply server/text_generation_server/models/causal_lm.py]: Done.

**Comment by mandy-li** [comment]
@yuanwu2017 , pls test if any performance regression for llama2, llama3.1, lava-next with this PR

**Comment by yuanwu2017** [comment]
@regisss  Please help to review the patch. We are preparing the test report and will send it to you later.

**Comment by yuanwu2017** [comment]
@regisss I ran the performance test of 2.3.1 combined with 2.0.6 and found no performance regression. Please help to review the patch
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/yuanwu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/yuanwu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


BF16 | cards | 2.3.1(Throughput   tokens/sec) | 2.0.6 |  
-- | -- | -- | -- | --
  |   | round1 | round2 | round3 | throughput  average | round1 | round2 | round3 | throughput  average | 2.3.1/2.0.,6
meta-llama/Llama-2-7b-chat-hf | 1X | 922.5504 | 916.2069 | 896.8379 | 911.8651 | 889.6319 | 937.6007 | 942.0372 | 923.0899 | 0.98784
meta-llama/Llama-2-7b-chat-hf | 8X | 1059.256 | 1093.806 | 1100.997 | 1084.686 | 1086.86 | 1092.776 | 1099.432 | 1093.023 | 0.992373
meta-llama/Llama-2-70b-chat-hf | 8X | 1560.617 | 1568.882 | 1565.688 | 1565.062 | 1543.909 | 1549.759 | 1554.364 | 1549.344 | 1.010145
meta-llama/Meta-Llama-3-8B-Instruct | 1X | 1656.585 | 1643.067 | 1549.219 | 1616.29 | 1653.716 | 1644.518 | 1656.652 | 1651.629 | 0.978604
meta-llama/Meta-Llama-3-8B-Instruct | 8X | 1312.84 | 1312.841 | 1301.689 | 1309.123 | 1300.603 | 1332.732 | 1362.129 | 1331.821 | 0.982957
meta-llama/Meta-Llama-3-70B-Instruct | 8X | 1857.385 | 1869.571 | 1867.97 | 1864.976 | 1794.821 | 1822.224 | 1798.287 | 1805.111 | 1.033164
meta-llama/Meta-Llama-3.1-8B-Instruct | 1X | 1665.54 | 1657.52 | 1656.056 | 1659.705 | 1681.042 | 1683.532 | 1676.813 | 1680.462 | 0.987648
meta-llama/Meta-Llama-3.1-8B-Instruct | 8X | 1327.326 | 1333.611 | 1339.457 | 1333.465 | 1327.424 | 1343.585 | 1349.487 | 1340.166 | 0.995
meta-llama/Meta-Llama-3.1-70B-Instruct | 8X | 1135.02 | 1119.769 | 1307.884 | 1187.557 | 888.1208 | 1517.145 | 1521.042 | 1125.552 | 1.055089
bigcode/starcoder2-3b | 1X | 1395.299 | 1399.807 | 1412.589 | 1402.565 | 1570.58 | 1424.191 | 1395.299 | 1463.357 | 0.958458
bigcode/starcoder2-3b | 2X | 807.3332 | 847.9122 | 848.0986 | 834.448 | 870.8548 | 815.4062 | 841.9448 | 842.7353 | 0.990166
bigcode/starcoder | 1X | 903.9525 | 950.4662 | 937.134 | 930.5176 | 896.6724 | 967.4937 | 941.3295 | 935.1652 | 0.99503
bigcode/starcoder | 8X | 742.8011 | 752.5386 | 759.0889 | 751.4762 |   |   |   | 0 | #DIV/0!
llava-hf/llava-v1.6-mistral-7b-hf | 1X | 1.21571 | 1.305719 | 1.309418 | 1.276949 | 1.260808 | 1.257572 | 1.261474 | 1.259951 | 1.013491
llava-hf/llava-v1.6-mistral-7b-hf | 8X |   |   |   | 0 |   |   |   | 0 | #DIV/0!
Mistral-7B-Instruct-v0.2 | 1X | 30.64058 | 31.18337 | 31.395 | 31.07298 | 30.79294 | 31.31473 | 31.45726 | 31.18831 | 0.996302
Mistral-7B-Instruct-v0.2 | 2X | 40.77183 | 42.09491 | 42.76818 | 41.87831 | 40.21588 | 42.83104 | 44.37624 | 42.47439 | 0.985966
mistralai/Mixtral-8x7B-Instruct-v0.1 | 1X | 54.70279 | 56.0435 | 56.12848 | 55.62492 | 20.55598 | 20.89971 | 20.93321 | 20.7963 | 2.674751
mistralai/Mixtral-8x7B-Instruct-v0.1 | 2X | 36.85146 | 39.75177 | 43.14368 | 39.91564 | 18.40397 | 18.65196 | 17.65057 | 18.2355 | 2.188897
FP8 | cards | 2.3.1(Throughput tokens/sec) |   | 2.0.6 |   |  
meta-llama/Llama-2-70b-chat-hf | 8X | 1374.273 | 1407.445 | 1392.253 | 1391.324 | 1406.499 | 1420.672 | 1392.096 | 1406.423 | 0.989264
meta-llama/Meta-Llama-3-70B-Instruct | 8X | 1653.407 | 1553.882 | 1548.01 | 1585.099 | 1527.075 | 1589.907 | 1633.097 | 1583.36 | 1.001099



</body>

</html>

**Comment by yuanwu2017** [comment]
@regisss @yao-matrix  I have ran the performance benchmark of v2.0.6 and v.2.3.1, and not found the performance regression. Please help to review the patch.
image

**Comment by yao-matrix** [comment]
@yuanwu2017 , pls resolve branch conflicts, thx.

**Comment by yao-matrix** [comment]
@yuanwu2017 , pls resolve branch conflicts, thx.

**Review by yao-matrix — APPROVED** [review]

**Review by regisss — APPROVED** [review]
LGTM!

**Pull request merged**